File: questions/q9/solution.py
import sys
import heapq

# ============================================
# IMPLEMENT YOUR SOLUTION IN THE main() FUNCTION
# ============================================

def main(lines):
    """
    Merge k sorted lists into a single sorted list.
    
    Args:
        lines: A list of strings where:
               - lines[0]: String containing k (number of lists)
               - lines[1] to lines[k]: Strings with space-separated integers
                 (or empty strings for empty lists)
    
    Your solution should print the merged sorted list to stdout
    """
    # ============================================
    # YOUR CODE HERE
    # ============================================
    
    k = int(lines[0])
    
    lists = []
    for i in range(1, k + 1):
        if lines[i].strip():
            lists.append(list(map(int, lines[i].split())))
        else:
            lists.append([])

    heap = []
    
    for i in range(k):
        if lists[i]:
            heapq.heappush(heap, (lists[i][0], i, 0))
    
    result = []
    
    while heap:
        value, list_index, element_index = heapq.heappop(heap)
        result.append(value)
        
        next_index = element_index + 1
        if next_index < len(lists[list_index]):
            heapq.heappush(heap, 
                (lists[list_index][next_index], list_index, next_index))

    print(" ".join(map(str, result)))
    
    # ============================================
    # END YOUR CODE
    # ============================================


# ============================================
# DO NOT MODIFY BELOW THIS LINE
# ============================================
if __name__ == "__main__":
    input_string = sys.stdin.read()
    lines = input_string.strip().split("\n")
    main(lines)

File: questions/q9/Solution.java
package q9;

import java.util.*;

// ============================================
// IMPLEMENT YOUR SOLUTION IN THE solve() METHOD
// ============================================

public class Solution {
    /**
     * Merge k sorted lists into a single sorted list.
     * 
     * @param lists A list of sorted integer lists
     * 
     * Your solution should print the merged sorted list to stdout
     */
    public static void solve(List<List<Integer>> lists) {
        // ============================================
        // YOUR CODE HERE
        // ============================================
        
        int k = lists.size();
        
        // Min heap: [value, list_index, element_index]
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        
        for (int i = 0; i < k; i++) {
            if (!lists.get(i).isEmpty()) {
                heap.offer(new int[]{lists.get(i).get(0), i, 0});
            }
        }
        
        List<Integer> result = new ArrayList<>();
        
        while (!heap.isEmpty()) {
            int[] entry = heap.poll();
            int value = entry[0];
            int listIndex = entry[1];
            int elemIndex = entry[2];
            
            result.add(value);
            
            int nextIndex = elemIndex + 1;
            if (nextIndex < lists.get(listIndex).size()) {
                heap.offer(new int[]{
                    lists.get(listIndex).get(nextIndex),
                    listIndex,
                    nextIndex
                });
            }
        }
        
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < result.size(); i++) {
            if (i > 0) sb.append(" ");
            sb.append(result.get(i));
        }
        System.out.println(sb.toString());
        
        // ============================================
        // END YOUR CODE
        // ============================================
    }
    
    // ============================================
    // DO NOT MODIFY BELOW THIS LINE
    // ============================================
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int k = Integer.parseInt(scanner.nextLine());
        
        List<List<Integer>> lists = new ArrayList<>();
        for (int i = 0; i < k; i++) {
            String line = scanner.nextLine();
            List<Integer> list = new ArrayList<>();
            if (!line.trim().isEmpty()) {
                String[] parts = line.trim().split("\\s+");
                for (String part : parts) {
                    list.add(Integer.parseInt(part));
                }
            }
            lists.add(list);
        }
        
        solve(lists);
        scanner.close();
    }
}